### PR TITLE
Add support for serialization of TR_BitVector

### PR DESCRIPTION
**Pull request by ashu-mehra**
Added methods for (de-)serialization of TR_BitVector.

Signed-off-by: Ashutosh Mehra <asmehra@redhat.com>

### CODE REVIEW

**Comment by ashu-mehra** [comment]
Changes in this PR are needed to support serialization of `TR_PersistentProfieInfo` in OpenJ9.
The OpenJ9 PR dependent on this is eclipse/openj9#9590

**Comment by ashu-mehra** [comment]
@Leonardo2718 @dsouzai - fyi

File: compiler/infra/BitVector.hpp [hunk 1]
```diff
@@ -163,6 +163,28 @@ class TR_BitVector
...
+      _lastChunkWithNonZero = *(int32_t *)buffer;
+      buffer += sizeof(_lastChunkWithNonZero);
+      _numChunks = *(int32_t *)buffer;
+      _chunks = NULL;
+      if (_numChunks > 0)
+         {
+         _chunks = (chunk_t*) TR_Memory::jitPersistentAlloc(_numChunks * sizeof(*_chunks), TR_Memory::BitVector);
+         memcpy(_chunks, buffer, _numChunks * sizeof(_chunks[0]));
+         buffer += (_numChunks * sizeof(_chunks[0]));
+         }
```
Comment (dsouzai) [inline compiler/infra/BitVector.hpp:183]: It would be better to use a struct for these fields, as you don't have to worry about getting the order exactly correct nor having to deal with any subtleties. Something like:

```
struct TR_SerializedBitVector
   {
   int32_t _firstChunkWithNonZero;
   int32_t _lastChunkWithNonZero;
   int32_t _numChunks;
   }
...
   TR_BitVector(uint8_t * &buffer)
      {
      TR_SerializedBitVector *data = (TR_SerializedBitVector *)buffer;
      _firstChunkWithNonZero = data->_firstChunkWithNonZero;
      _lastChunkWithNonZero = data->_lastChunkWithNonZero;
      _numChunks = data->_numChunks;
     ...
     uint8_t *serializedChunks = buffer + sizeof(TR_SerializedBitVector);
     memcpy(_chunks, serializedChunks, _numChunks * sizeof(_chunks[0]));
     buffer = serializedChunks + (_numChunks * sizeof(_chunks[0]));
     ...
```
Also, put `_numChunks * sizeof(*_chunks)` into a variable since it's used three times.

File: compiler/infra/BitVector.hpp [hunk 2]
```diff
@@ -838,6 +860,42 @@ class TR_BitVector
...
+      buffer += sizeof(_firstChunkWithNonZero);
+      *(int32_t *)buffer = _lastChunkWithNonZero;
+      buffer += sizeof(_lastChunkWithNonZero);
+      *(int32_t *)buffer = _numChunks;
+      buffer += sizeof(_numChunks);
+      if (_numChunks > 0)
+         {
+         memcpy(buffer, _chunks, _numChunks * sizeof(_chunks[0]));
+         buffer += (_numChunks * sizeof(_chunks[0]));
+         }
```
Comment (dsouzai) [inline compiler/infra/BitVector.hpp:896]: Same comment as above.

**Review by dsouzai — CHANGES_REQUESTED** [review]

**Review by dsouzai — APPROVED** [review]

File: compiler/infra/BitVector.hpp [hunk 3]
```diff
@@ -838,6 +874,40 @@ class TR_BitVector
...
+         size += (_numChunks * sizeof(*_chunks));
+         }
+      return size;
+      }
+
+   /**
+    * @brief Serialize this object
+    * @param [in] serializer used for serializing this object
+    */
+   void serialize(uint8_t * &buffer) const
```
Comment (fjeremic) [inline compiler/infra/BitVector.hpp:896]: Is this a standardized signature across OMR? I'm not quite understanding why this is not a `void*` given we cast the type anyway.

Comment (ashu-mehra) [reply compiler/infra/BitVector.hpp:896]: @fjeremic `uint8_t *` makes it easier to increment the buffer and I want the buffer to be updated which is why I made it a reference. But I can instead use the following signature:
`void *serialize(void *buffer) const`
This would return the updated buffer. Does that look ok?

Comment (fjeremic) [reply compiler/infra/BitVector.hpp:896]: No this is fine. I think we should at the very least document that `buffer` will get mutated as data is written and that it is up to the user to determine that there is enough space available.

Comment (ashu-mehra) [reply compiler/infra/BitVector.hpp:900]: @fjeremic Updated the comment as suggested.

File: compiler/infra/BitVector.hpp [hunk 4]
```diff
@@ -838,6 +874,44 @@ class TR_BitVector
...
+      {
+      TR_SerializedBitVector *sbv = (TR_SerializedBitVector *)buffer;
+      sbv->_firstChunkWithNonZero = _firstChunkWithNonZero;
+      sbv->_lastChunkWithNonZero = _lastChunkWithNonZero;
+      sbv->_numChunks = _numChunks;
+      buffer += sizeof(TR_SerializedBitVector);
+      if (_numChunks > 0)
+         {
+         uint32_t chunksSize = _numChunks * sizeof(*_chunks);
+         memcpy(buffer, _chunks, chunksSize);
```
Comment (fjeremic) [inline compiler/infra/BitVector.hpp:910]: `chunksSize` should be of type `size_t` since this is what `memcpy` accepts.

File: compiler/infra/BitVector.hpp [hunk 5]
```diff
@@ -838,6 +874,44 @@ class TR_BitVector
...
+    * @brief Serialize this object
+    * @param [in] serializer used for serializing this object
+    *
+    * Note that the caller should ensure buffer is large enough to accommodate the serialized data.
+    * @see getSizeForSerialization()
+    * On return the buffer gets updated to point to the location past the serialized data.
+    */
+   void serialize(uint8_t * &buffer) const
+      {
+      TR_SerializedBitVector *sbv = (TR_SerializedBitVector *)buffer;
```
Comment (fjeremic) [inline compiler/infra/BitVector.hpp:902]: We should use `reinterpret_cast` here. Similarly above.

File: compiler/infra/BitVector.hpp [hunk 6]
```diff
@@ -838,6 +874,44 @@ class TR_BitVector
...
+      }
+
+   /**
+    * @brief Serialize this object
+    * @param [in] serializer used for serializing this object
+    *
+    * Note that the caller should ensure buffer is large enough to accommodate the serialized data.
+    * @see getSizeForSerialization()
+    * On return the buffer gets updated to point to the location past the serialized data.
+    */
```
Comment (fjeremic) [inline compiler/infra/BitVector.hpp:899]: This needs a `@note` Doxygen tag.

File: compiler/infra/BitVector.hpp [hunk 7]
```diff
@@ -163,6 +173,32 @@ class TR_BitVector
    TR_BitVector() : _numChunks(0), _chunks(NULL), _firstChunkWithNonZero(0), _lastChunkWithNonZero(-1), _growable(growable), _region(0) { }
    TR_BitVector(TR::Region &region) : _numChunks(0), _chunks(NULL), _firstChunkWithNonZero(0), _lastChunkWithNonZero(-1), _growable(growable), _region(&region) { }
 
+   /**
+    * @brief Constructor to create a new BitVector by reading serialized data from the memory buffer
+    * @param [in] buffer Memory buffer containing serialized BitVector
+    */
+   TR_BitVector(uint8_t * &buffer)
```
Comment (fjeremic) [inline compiler/infra/BitVector.hpp:180]: This needs similar documentation to `serialize` to note how much data should be read. This constructor implicitly assumes there is a certain amount of data present and it reads memory unconditionally.

**Comment by dsouzai** [comment]
You know, if you're going to change the code up anyway, a minor nitpick I missed earlier is that `TR_SerializedBitVector` could just be an inner struct in `TR_BitVector`, in which case it would probably also best renamed  just `Serialized` (I could imagine in the future some form of templatized function taking in a `T::Serialized` object...though maybe this is starting to border on overengineering heh, so do what you think is best). If you really wanted it accessible outside, you could make it public.

**Comment by ashu-mehra** [comment]
Pushed the changes based on above suggestions.

File: cmake/modules/platform/toolcfg/gnu.cmake [hunk 8]
```diff
@@ -25,7 +25,7 @@ set(OMR_ENHANCED_WARNING_FLAG -Wall)
 
 # disable builtin strncpy buffer length check for components that use variable length
 # array fields at the end of structs
-set(OMR_STRNCPY_FORTIFY_OPTIONS -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1)
+set(OMR_STRNCPY_FORTIFY_OPTIONS -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=0)
```
Comment (fjeremic) [inline cmake/modules/platform/toolcfg/gnu.cmake:28]: Was this change intended?

Comment (ashu-mehra) [reply cmake/modules/platform/toolcfg/gnu.cmake:28]: oops! I will remove that.

**Review by fjeremic — APPROVED** [review]

**Comment by fjeremic** [comment]
@genie-omr build all
